tools/autograd/gen_python_functions.py

```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
 
     # first group by signature ignoring out arguments
-    for declaration in declarations:
-        signature = get_python_signature(declaration, is_python_method, skip_outputs=True)
-        v = grouped[signature]
-        if declaration['name'].endswith('_out'):
-            v['out'] = declaration
-            # prefer the signature with optional out=... arguments
-            v['signature'] = get_python_signature(declaration, is_python_method)
+    for overload in overloads:
+        sig = overload.signature.signature_str(skip_outputs=True)
+        if overload.function.func.is_out_fn():
+            outplaces[sig] = overload
         else:
-            v['base'] = declaration
-            if 'signature' not in v:
-                v['signature'] = signature
-
-    result = []
-    for x, dictionary in sorted(grouped.items()):
-        if 'base' not in dictionary:
-            candidates = []
-            non_out_name = dictionary['out']['operator_name']
-            for declaration in declarations:
-                if declaration['name'] == non_out_name and not declaration['deprecated']:
-                    signature = get_python_signature(declaration, is_python_method, skip_outputs=True)
-                    candidates.append(signature)
+            bases[sig] = overload
+
+    for sig, out in outplaces.items():
+        if sig not in bases:
+            candidates: List[str] = []
+            for overload in overloads:
+                if str(overload.function.func.name.name) == str(out.function.func.name.name) \
+                        and not overload.function.func.is_out_fn() \
+                        and not overload.signature.deprecated:
+                    candidates.append(overload.signature.signature_str(skip_outputs=True))
+            out_sig = out.signature.signature_str()
             raise RuntimeError(
-                "While identifying overloads, we found an out schema {} without a corresponding non-out variant. "
-                "We expected the non-out variant to have schema: \n- {}\nPlease check that you spelled the schema "
+                f"While identifying overloads, we found an out schema {out_sig} without a corresponding non-out variant. "
+                f"We expected the non-out variant to have schema: \n- {sig}\nPlease check that you spelled the schema "
                 "correctly in native_functions.yaml. We discovered the following candidate(s): \n"
-                .format(dictionary['signature'], x) + "\n".join("- {}".format(candidate) for candidate in candidates))
-        result.append(dictionary)
-    return sort_declarations(result)
+                + "\n".join(f"- {candidate}" for candidate in candidates))
+
+    grouped: List[PythonSignatureGroup] = []
+    for sig, base in bases.items():
+        outplace = outplaces.get(sig)
+        grouped.append(PythonSignatureGroup(
+            # prefer the signature with optional out=... arguments
```

Comment:
Nothing to do here, but I hate this. Will be kind of hard to fix though.

The opposite invariant (use the signature with functional arguments) would be easier to get to; I doubt there are many out= functions that don't have a functional version. Something that isn't so clear just looking at the comment here or in PythonSignatureGroup comments is *why* we need to prefer the out= variant.